app/cni/pkg/cni/kubernetes_linux.go

```diff
@@ -2,41 +2,108 @@ package cni
 
 import (
 	"context"
+	"slices"
 
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/sethvargo/go-retry"
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+	k8s_util "github.com/kumahq/kuma/pkg/plugins/runtime/k8s/util"
 )
 
-func newKubeClient(conf PluginConf) (*kubernetes.Clientset, error) {
-	kubeconfig := conf.Kubernetes.Kubeconfig
+func newKubeClient(logger logr.Logger, conf PluginConf) (*kubernetes.Clientset, error) {
+	logger = logger.WithValues("kubeconfig", conf.Kubernetes.Kubeconfig)
+
 	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
-		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfig},
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: conf.Kubernetes.Kubeconfig},
 		&clientcmd.ConfigOverrides{},
 	).ClientConfig()
 	if err != nil {
-		log.Error(err, "failed setting up kubernetes client with kubeconfig", "kubeconfig", kubeconfig)
+		logger.Error(err, "failed setting up kubernetes client with kubeconfig")
 		return nil, err
 	}
 
-	log.V(1).Info("set up kubernetes client with kubeconfig", "kubeconfig", kubeconfig, "config", config)
+	logger.V(1).Info("set up kubernetes client with kubeconfig", "config", config)
 
 	return kubernetes.NewForConfig(config)
 }
 
-// getK8sPodInfo returns information of a POD
-func getKubePodInfo(ctx context.Context, client *kubernetes.Clientset, podName, podNamespace string) (int, map[string]struct{}, map[string]string, error) {
-	pod, err := client.CoreV1().Pods(podNamespace).Get(ctx, podName, metav1.GetOptions{})
-	log.V(1).Info("pod info", "pod", pod)
+func getAndValidatePodAnnotations(
+	ctx context.Context,
+	logger logr.Logger,
+	conf *PluginConf,
+	k8sArgs K8sArgs,
+) (map[string]string, bool, error) {
+	name := string(k8sArgs.K8S_POD_NAME)
+	namespace := string(k8sArgs.K8S_POD_NAMESPACE)
+
+	if namespace == "" || name == "" {
+		return nil, true, errors.New("pod namespace or pod name is empty")
+	}
+
+	if slices.Contains(conf.Kubernetes.ExcludeNamespaces, namespace) {
+		return nil, true, errors.New("namespace is in the exclude list")
+	}
+
+	client, err := newKubeClient(logger, *conf)
 	if err != nil {
-		log.Error(err, "can't get pod info")
-		return 0, nil, nil, err
+		return nil, false, errors.Wrap(err, "failed to create Kubernetes client")
+	}
+
+	var pod *corev1.Pod
+	if err := retry.Do(
+		ctx,
+		retry.WithMaxRetries(podRetrievalMaxRetries, retry.NewConstant(podRetrievalInterval)), // backoff
```

Comment:
so is the backoff incoming?

Reply:
it's backoff value for retries, but I have placed it in separate variable to be clearer